docs/proposals/vault-auth-provider.rst

```diff
@@ -0,0 +1,154 @@
+
+  - This should have a unique name based on the AWS Instance ID / boot ID / etc
+- Like any other auth provider, ultimately make and return a vault-token (from a template?), including the unique vault-policy just made
+
+Changes
+=======
+
+Configuration
+-------------
+
+Config of the provider will be though Vault paths as normal.
+
+``config/client``
+~~~~~~~~~~~~~~~~~
+
+Global configuration of the plugin.
+
+Fields:
+
+::
+
+    cloud_provider=[aws|gce]
+    aws_access_id= (use instance role if empty)
+    aws_secret_id= (use instance role if emtpy)
+    aws_region=    (detect from metadata service if empty)
+    gce_*          (equivalents...)
+
+``roles/<rolename>``
+~~~~~~~~~~~~~~~~~~~~
+
+Tell Vault about the IAM roles in use by the instances.
+
+Recall that each instance we bring up has an IAM role attached depending on its type, e.g. etcd, master, or worker. We can easily use this to tell the different instance types apart, as they need different policy templates.
+
+Note: we have to store this information at a vault path, which means coming up with yet another set of symbolic names, and a name for this type of thing. I'll loosely call them "roles", as there's a 1:1 mapping with the IAM Roles they're modelling. Recall that the vault server is shared between clusters. We won't do any namespacing in the path, so we should strongly encourage (or enforce?) that names are e.g. alice\_cluster-etcd\_role
+
+Fields:
+
+::
+
+    iam_role_arn="arn:aws:iam::$account:role/$role"
+    base_path="..."  # base path to cluster's vault secrets, such that the kubernetes PKI  lives at {{base_path}}/pki/k8s. E.g. "dev-cluster" 
+
+``templates/<rolename>/<templatename>``
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Recall that we need to dynamically create two types of thing from templates:
+
+- vault-policies; to attach to tokens, which amoungst other things point at pki-roles
+- pki-roles; to actually limit EC2 Instances' PKI power. These are stored as generic secrets
+
+Semantics:
+
+- These templates are to be golang templates, and substitute at least: ``{{.InstanceHash}}``, ``{{.FQDN}}``, ``{{.InternalIPv4}}``, ``{{.BasePath}}``
+- Secrets are to be specified in JSON \* Policies are to be specified in vault free-edition policy-language 
+- ``path`` is where the rendered template should be written during the log-in process, relative to the ``base_path``
+
+  - e.g. /pki/k8s/roles/:name would result in a pki-role at /alice\_cluster/pki/k8s/roles/kubelet
+  - e.g. such a role for the kubelet would be crated by a "worker" role.
+
+Fields:
+
+::
+
+    type="policy|generic"
+    path="relative/path/of/template/output"
+    template="<golang template of either policy document or JSON-encoded generic secret>"
+
+
+Notable items
+=============
+
+Concerns
+--------
+
+-  Huge part of security critical code in our hands
+-  Clean up of roles and templates once they are no longer used
```

Comment:
We will also need to extend the vault auth plugin each time we expand tarmak to support a new cloud, or produce a `tarmak-auth-aws`, `tarmak-auth-gcp`, `tarmak-auth-vsphere`, `tarmak-auth-token` etc plugin.

Reply:
This is true. It think I would rather have a single provider with a different release cycle than tarmak itself (maybe as part of vault-helper, which needs a better name). But IMHO a single plugin binary should support mutliple providers. (See the config endpioint flags that are mounted per instance of the plugin)

Reply:
Fair point, added to the document